src/polydb.jl

```diff
@@ -0,0 +1,301 @@
+module Polydb
+
+import Polymake: call_function
+
+using Polymake
+
+# julia itself also has a cert.pem but this one should be more recent
+# and provides a variable for the path
+using MozillaCACerts_jll
+
+using Mongoc
+
+#Polymake.Polydb's types store information via
+# a corresponding Mongoc type variable
+struct Database
+   mdb::Mongoc.Database
+end
+
+struct Collection{T}
+   mcol::Mongoc.Collection
+end
+
+struct Cursor{T}
+   mcursor::Mongoc.Cursor{Mongoc.Collection}
+end
+
+# connects to the Polydb and
+# returns a Polymake.Polydb.Database instance
+function get_db()
+   # we explicitly set the cacert file, otherwise we might get connection errors because the certificate cannot be validated
+   client = Mongoc.Client("mongodb://polymake:database@db.polymake.org/?authSource=admin&ssl=true&sslCertificateAuthorityFile=$(cacert)")
+   return Database(client["polydb"])
+end
+
+# returns a Polymake.Polydb.Collection instance with the given name
+# sections and collections in the name are connected with the '.' sign,
+# i.e. names = "Polytopes.Lattice.SmoothReflexive"
+function get_collection(db::Database, name::String)
+   return Collection{Polymake.BigObject}(db.mdb[name])
+end
+
+function get_collection(db::Database, name::String, ::Type{T}) where T <: Union{Polymake.BigObject, Mongoc.BSON}
+   return Collection{T}(db.mdb[name])
+end
+
+# search a collection for documents matching the criteria given by d
+function find(c::Collection{T}, d::Dict=Dict(); opts::Union{Nothing, Dict}=nothing) where T
```

Comment:
why not `Mongoc.find` ?

Reply:
here we define the `Polymake.Polydb.find` function, calling `Mongoc.find` in the background. for general purposes in context of the database, the user does not need to `using Mongoc`. The only case this does not apply to is handling the `Mongoc.BSON` templated types, forcing the user to `using Mongoc`. But we could work around that by using another parameter instead of the type; access to entries of `Mongoc.BSON` variables received from polymake still works without `using Mongoc` (i guess because the corresponding `Base.index` method is kept)

Reply:
my question was rather: since the first argument is `Polymake.Polydb.Collection` why do we chose not to add a method to `Mongoc.find` and rather define a separate `Polymake.Polydb.find` function?

Is there a conflict with signatures?

Reply:
the idea behind this is that the user does not have to call `using Mongoc` to access core features regarding polyDB. i would consider the `find` method to be significant enough.
of course we could change the method to `Mongoc.find`, but do we gain anything from it?

Reply:
If You write here `Mongoc.find` You're adding a new overloaded method that sits inside `Polydb` module. It just extends the other `find`, users will have acces to `Mongoc.find` without specifying the parent module if they're `using Polydb: find`

Reply:
i changed it to `Mongoc.find`, but even when `importing Mongoc` instead of `using` in line 5, i encounter the following problem if i try to use it in a session:

```julia
julia> results1 = Mongoc.find(collection, query1)
ERROR: UndefVarError: Mongoc not defined
Stacktrace:
 [1] top-level scope at none:0

julia> results1 = Polymake.Polydb.find(collection, query1)
ERROR: UndefVarError: find not defined
Stacktrace:
 [1] getproperty(::Module, ::Symbol) at ./Base.jl:26
 [2] top-level scope at none:0

julia> using Mongoc

julia> results1 = Mongoc.find(collection, query1)
Polymake.Polydb.Cursor{Polymake.BigObject}(Mongoc.Cursor{Mongoc.Collection}(Collection(Database(Client(URI("mongodb://polymake:database@db.polymake.org/?authSource=admin&ssl=true&sslCertificateAuthorityFile=/home/jordan/.julia/artifacts/59cb7446ab8874d1c1c4df2eee3cba7c76ed4d1d/share/cacert.pem")), "polydb"), "Polytopes.Lattice.SmoothReflexive"), Ptr{Nothing} @0x000000000c9c1f00))
```

what am i missing?

Reply:
You need to `import Mongoc: find`
when you add new methods to a function it's not automatically brought into scope
```julia
julia> methods(Polymake.Polydb.find)
# 9 methods for generic function "find":
[1] find(bucket::Mongoc.Bucket, bson_filter::Mongoc.BSON; options) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/gridfs.jl:172
[2] find(bucket::Mongoc.Bucket) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/gridfs.jl:172
[3] find(collection::Mongoc.CollectionSession, bson_filter::Mongoc.BSON; options) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/session.jl:93
[4] find(collection::Mongoc.CollectionSession) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/session.jl:93
[5] find(collection::Mongoc.Collection, bson_filter::Mongoc.BSON; options) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/collection.jl:203
[6] find(collection::Mongoc.Collection) in Mongoc at /home/kalmar/.julia/packages/Mongoc/OxMdy/src/collection.jl:203
[7] find(c::Polymake.Polydb.Collection{T}) where T in Polymake.Polydb at /home/kalmar/.julia/dev/Polymake/src/polydb.jl:43
[8] find(c::Polymake.Polydb.Collection{T}, d::Dict; opts) where T in Polymake.Polydb at /home/kalmar/.julia/dev/Polymake/src/polydb.jl:43
[9] find(c::Polymake.Polydb.Collection{T}, d::Pair...) where T in Polymake.Polydb at /home/kalmar/.julia/dev/Polymake/src/polydb.jl:47
```